flink-ml-lib/src/main/java/org/apache/flink/ml/classification/linear/LogisticRegression.java

```diff
@@ -0,0 +1,653 @@
+import org.apache.flink.ml.param.Param;
+import org.apache.flink.ml.util.ParamUtils;
+import org.apache.flink.ml.util.ReadWriteUtils;
+import org.apache.flink.runtime.state.StateInitializationContext;
+import org.apache.flink.runtime.state.StateSnapshotContext;
+import org.apache.flink.streaming.api.datastream.DataStream;
+import org.apache.flink.streaming.api.datastream.SingleOutputStreamOperator;
+import org.apache.flink.streaming.api.environment.StreamExecutionEnvironment;
+import org.apache.flink.streaming.api.operators.AbstractStreamOperator;
+import org.apache.flink.streaming.api.operators.AbstractUdfStreamOperator;
+import org.apache.flink.streaming.api.operators.BoundedMultiInput;
+import org.apache.flink.streaming.api.operators.BoundedOneInput;
+import org.apache.flink.streaming.api.operators.OneInputStreamOperator;
+import org.apache.flink.streaming.api.operators.TwoInputStreamOperator;
+import org.apache.flink.streaming.runtime.streamrecord.StreamRecord;
+import org.apache.flink.table.api.Table;
+import org.apache.flink.table.api.bridge.java.StreamTableEnvironment;
+import org.apache.flink.table.api.internal.TableImpl;
+import org.apache.flink.util.Collector;
+import org.apache.flink.util.OutputTag;
+import org.apache.flink.util.Preconditions;
+
+import org.apache.commons.collections.IteratorUtils;
+
+import java.io.IOException;
+import java.util.ArrayList;
+import java.util.Arrays;
+import java.util.Collections;
+import java.util.HashMap;
+import java.util.List;
+import java.util.Map;
+import java.util.Random;
+
+/**
+ * This class implements methods to train a logistic regression model. For details, see
+ * https://en.wikipedia.org/wiki/Logistic_regression.
+ */
+public class LogisticRegression
+        implements Estimator<LogisticRegression, LogisticRegressionModel>,
+                LogisticRegressionParams<LogisticRegression> {
+
+    private Map<Param<?>, Object> paramMap = new HashMap<>();
+
+    private static final OutputTag<Tuple2<double[], double[]>> MODEL_OUTPUT =
+            new OutputTag<Tuple2<double[], double[]>>("MODEL_OUTPUT") {};
+
+    public LogisticRegression() {
+        ParamUtils.initializeMapWithDefaultValues(this.paramMap, this);
+    }
+
+    @Override
+    public Map<Param<?>, Object> getParamMap() {
+        return paramMap;
+    }
+
+    @Override
+    public void save(String path) throws IOException {
+        ReadWriteUtils.saveMetadata(this, path);
+    }
+
+    public static LogisticRegression load(StreamExecutionEnvironment env, String path)
+            throws IOException {
+        return ReadWriteUtils.loadStageParam(path);
+    }
+
+    @Override
+    @SuppressWarnings("unchecked")
+    public LogisticRegressionModel fit(Table... inputs) {
+        Preconditions.checkArgument(inputs.length == 1);
+        StreamTableEnvironment tEnv =
+                (StreamTableEnvironment) ((TableImpl) inputs[0]).getTableEnvironment();
+
+        DataStream<Tuple3<Double, Double, double[]>> trainData =
```

Comment:
The statement for calculating `Tuple3.of(...)` is a bit too long. And the code has extra `returns` statement which hopefully could be removed.

Would the following code be a bit simpler?

```
DataStream<Tuple3<Double, Double, double[]>> trainData =
        tEnv.toDataStream(inputs[0])
                .map(
                        new MapFunction<Row, Tuple3<Double, Double, double[]>>() {
                            @Override
                            public Tuple3<Double, Double, double[]> map(Row dataPoint) {
                                Double weight =
                                        getWeightCol() == null
                                                ? new Double(1.0)
                                                : (Double)
                                                        dataPoint.getField(getWeightCol());

                                return Tuple3.of(
                                        weight,
                                        (Double) dataPoint.getField(getLabelCol()),
                                        (double[]) dataPoint.getField(getFeaturesCol()));
                            }
                        });
```

Reply:
As discussed offline, we decided to go with a specific class `LabeledPointWithWeight`.